orientation-event: Add Permissions Policy integration tests

rakuco opened the pull request:
Add tests for deviceorientationabsolute, deviceorientation and devicemotion
as described in
https://w3c.github.io/deviceorientation/#permissions-policy-integration:

- The deviceorientation event requires the "accelerometer" and "gyroscope"
  features.
- The deviceorientationabsolute event requires the "accelerometer",
  "gyroscope" and "magnetometer" features.
- The devicemotion event requires the "accelerometer" and "gyroscope"
  features.

(WebKit currently requires the "magnetometer" feature for deviceorientation
events, see https://bugs.webkit.org/show_bug.cgi?id=271891).

The tests being added follow permissions-policy/README.md's recommendations:
- Tests with different Permissions-Policy header values and how they
  influence usage in the top-level frame, same-origin iframes and
  cross-origin iframes.
- Different values for the "allow" attribute and how they influece
  same-origin and cross-origin iframes.
- Usage of the "allow" attribute with redirection.

This is the minimum amount of coverage we need. These tests can be expanded
to include:
- Calls to `run_all_fp_tests_allow_self()` to test the behavior of the
  default policy for each feature (it requires making the function work with
  more than one feature).
- The relative orientation to absolute orientation fallback behavior that
  depends on whether the "magnetometer" feature is enabled.
- Not enabling all features (it may cause an explosion in the number of
  tests though).

Commit:
orientation-event: Add Permissions Policy integration tests

Add tests for deviceorientationabsolute, deviceorientation and devicemotion
as described in
https://w3c.github.io/deviceorientation/#permissions-policy-integration:

- The deviceorientation event requires the "accelerometer" and "gyroscope"
  features.
- The deviceorientationabsolute event requires the "accelerometer",
  "gyroscope" and "magnetometer" features.
- The devicemotion event requires the "accelerometer" and "gyroscope"
  features.

(WebKit currently requires the "magnetometer" feature for deviceorientation
events, see https://bugs.webkit.org/show_bug.cgi?id=271891).

The tests being added follow permissions-policy/README.md's recommendations:
- Tests with different Permissions-Policy header values and how they
  influence usage in the top-level frame, same-origin iframes and
  cross-origin iframes.
- Different values for the "allow" attribute and how they influece
  same-origin and cross-origin iframes.
- Usage of the "allow" attribute with redirection.

This is the minimum amount of coverage we need. These tests can be expanded
to include:
- Calls to `run_all_fp_tests_allow_self()` to test the behavior of the
  default policy for each feature (it requires making the function work with
  more than one feature).
- The relative orientation to absolute orientation fallback behavior that
  depends on whether the "magnetometer" feature is enabled.
- Not enabling all features (it may cause an explosion in the number of
  tests though).

reillyeon commented on orientation-event/orientation/permissions-policy/deviceorientationabsolute-allowed-by-permissions-policy-attribute-redirect-on-load.https.html:
```diff
@@ -0,0 +1,53 @@
+    'NO-EVENT',
+    "accelerometer; gyroscope; magnetometer",
+  );
+}, `"allow" disallows allows cross-origin navigantion in an iframe`);
```
```suggestion
}, `"allow" disallows cross-origin navigation in an iframe`);
```

reillyeon commented on orientation-event/orientation/permissions-policy/deviceorientation-allowed-by-permissions-policy-attribute-redirect-on-load.https.html:
```diff
@@ -0,0 +1,53 @@
+    JSON.stringify(orientationData),
+    "accelerometer; gyroscope",
+  );
+}, `"allow" attribute allows same-origin navigantion in an iframe`);
```
```suggestion
}, `"allow" attribute allows same-origin navigation in an iframe`);
```

reillyeon commented on orientation-event/orientation/permissions-policy/deviceorientation-allowed-by-permissions-policy-attribute-redirect-on-load.https.html:
```diff
@@ -0,0 +1,53 @@
+    'NO-EVENT',
+    "accelerometer; gyroscope",
+  );
+}, `"allow" disallows allows cross-origin navigantion in an iframe`);
```
```suggestion
}, `"allow" disallows cross-origin navigation in an iframe`);
```

reillyeon reviewed (approved):
LGTM with nits

marcoscaceres commented on orientation-event/motion/permissions-policy/devicemotion-allowed-by-permissions-policy-attribute-redirect-on-load.https.html:
```diff
@@ -0,0 +1,57 @@
+  await helper.initializeSensors();
+
+  const motionData = generateMotionData(1, 2, 3,
+                                        4, 5, 6,
```
Nit: can we run prettier over this? Otherwise we use up with inconsistent/arbitrary whitespace usage and closing brackets etc.

rakuco replied:
Do you mean [this prettier](https://prettier.io/)? I'm fine with doing that, I just can't find any references to it in the tree or which settings should be used (I see a lot of tunables in https://prettier.io/playground/).

marcoscaceres replied:
I just use the defaults when I run it in Visual Studio Code.

rakuco replied:
I've run prettier on all files here. PTAL.

marcoscaceres commented on orientation-event/motion/permissions-policy/devicemotion-enabled-on-self-origin-by-permissions-policy.https.html:
```diff
@@ -0,0 +1,69 @@
+promise_test(async t => {
+  const helper = new SensorTestHelper(t, 'devicemotion');
+  await helper.grantSensorsPermissions();
+  await helper.initializeSensors();
```
This puts a dependency on create_virtual_sensor :( Are we sure there isn't some other way to test this? 

With other permission policy tests, we usually just have a simple iframe that tries to use the API.

rakuco replied:
> This puts a dependency on create_virtual_sensor :( Are we sure there isn't some other way to test this?

I don't think so. This API does not have any mechanisms to throw an exception when permissions policy checks fail like other APIs do, so we need a way to verify that the checks have failed, in which case we wait for readings _not_ to be delivered. We could try to reach consensus on w3c/deviceorientation#118, but it wouldn't help with the case in which we need to verify that the checks do pass, as without virtual sensors we'd be entirely dependent on the presence of specific hardware for them to work.

marcoscaceres replied:
Ok, yeah... I was trying to think of something too... and the only thing was trying to create the events in JS, which is silly.

Commit:
Fix typos

Co-authored-by: Reilly Grant <reillyeon@users.noreply.github.com>

Commit:
Fix typos

Co-authored-by: Reilly Grant <reillyeon@users.noreply.github.com>

Commit:
Fix typos

Co-authored-by: Reilly Grant <reillyeon@users.noreply.github.com>

Commit:
Fix typos

Co-authored-by: Reilly Grant <reillyeon@users.noreply.github.com>

Commit:
Fix typos

Co-authored-by: Reilly Grant <reillyeon@users.noreply.github.com>

Commit:
Fix typos

Co-authored-by: Reilly Grant <reillyeon@users.noreply.github.com>

marcoscaceres commented on permissions-policy/resources/permissions-policy-orientation-event.sub.html:
```diff
@@ -0,0 +1,13 @@
+<!DOCTYPE html>
```
Nit: this file doesn't need `.sub.` by the looks of it.

marcoscaceres commented on permissions-policy/resources/permissions-policy-orientation-event.sub.html:
```diff
@@ -0,0 +1,13 @@
+<!DOCTYPE html>
+<body>
+<script src="/orientation-event/resources/orientation-event-helpers.js"></script>
+<script src="orientation-event.js"></script>
```
nit

marcoscaceres commented on permissions-policy/resources/permissions-policy-orientation-event.sub.html:
```diff
@@ -0,0 +1,13 @@
+<script>
+"use strict";
+
+Promise.resolve().then(async () => {
+  const name = await waitForOrientationEvent('{{GET[eventName]}}');
+  window.parent.postMessage({ type: "availability-result", name }, "*");
+});
```
The Promise.resolve() doesn't seem to add much... Maybe just: 

```suggestion
waitForOrientationEvent('{{GET[eventName]}}').then(name => {
  window.parent.postMessage({ type: "availability-result", name }, "*");
});
```

rakuco replied:
The Promise.resolve() thing was copied from another file in `resources/`. It's not very intuitive, but I think it's equivalent to what other files do by running the code in `window.load`.

I've switched to adding a "load" event handler and replaced async/await with a regular Promise.then() flow like you suggested.

marcoscaceres commented on permissions-policy/resources/orientation-event.js:
```diff
@@ -0,0 +1,60 @@
+
+async function waitForOrientationEvent(eventName) {
+  if (eventName !== 'devicemotion' && eventName !== 'deviceorientation' && eventName !== 'deviceorientationabsolute') {
+    return 'ERROR';
```
I wonder if we should throw real errors here. It might make things easier to debug if things go wrong.

marcoscaceres commented on permissions-policy/resources/orientation-event.js:
```diff
@@ -0,0 +1,60 @@
+    // policy checks fail, so what we do here is wait for the timeout and the
+    // devicemotion event handler to race each other.
+    value = await new Promise((resolve, reject) => {
+      const timeoutId = window.setTimeout(() => {
```
This won't fly... it's gonna be super racy on testing infrastructure.

rakuco replied:
Do you have anything else in mind? This API doesn't throw any errors or fire any events when a permissions/permissions policy check fails, so the only other option would be adding a ReportingObserver and waiting for a [permissions-policy-violation report](https://w3c.github.io/webappsec-permissions-policy/#reporting). Would that be acceptable?

marcoscaceres commented on permissions-policy/resources/orientation-event.js:
```diff
@@ -0,0 +1,60 @@
+    value = await new Promise((resolve, reject) => {
+      const timeoutId = window.setTimeout(() => {
+        window.removeEventListener(eventName, handler);
+        reject('NO-EVENT');
```
Reject with real error if we do this... otherwise we lose the error stack.

rakuco replied:
Is not having an error stack really a problem in this case? `test_feature_availability_with_post_message_result()` expects a string, so throwing an Error here would essentially require some plumbing that would ultimately result in a string comparison anyway.

Commit:
Reformat files using the prettier VS Code extension

Commit:
Add a load event handler instead of relying on Promise.resolve()

The Promise.resolve() idea was copy-pasted from another file, but adding a
load event handler makes it more clear and in line with other files in
permissions-policy/resources.

While here, use Promise.then() instead of async/await to match other files
as well.

marcoscaceres commented:
Hey, promise to take a look soon! At a conference this week, but will try to sneak these in!
